src/ows/src/main/java/org/geoserver/ows/FileItemCleanupCallback.java

```diff
@@ -0,0 +1,45 @@
+/* (c) 2022 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.ows;
+
+import java.io.Reader;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.commons.fileupload.FileItem;
+import org.geotools.util.logging.Logging;
+
+/**
+ * Dispatcher callback to ensure that all uploaded files are deleted at the end of a
+ * multipart/form-data request instead of relying on the garbage collector to delete the files
+ * through the {@link org.apache.commons.fileupload.disk.DiskFileItem#finalize()} method.
+ */
+public class FileItemCleanupCallback extends AbstractDispatcherCallback {
+
+    private static final Logger LOGGER = Logging.getLogger(FileItemCleanupCallback.class);
+
+    private static final ThreadLocal<List<FileItem>> FILE_ITEMS =
+            ThreadLocal.withInitial(Collections::emptyList);
+
+    public static void setFileItems(List<FileItem> fileItems) {
+        FILE_ITEMS.set(fileItems);
+    }
+
+    @Override
+    public void finished(Request request) {
+        List<FileItem> items = FILE_ITEMS.get();
+        FILE_ITEMS.remove();
+        if (!items.isEmpty()) {
+            try (Reader r = request.getInput()) {
```

Comment:
Hmm... this input could also be coming from the HttpServletRequest.getInputStream(). I vaguely remember Tomcat getting real mad at us (like, hard segfault when using APR) for messing with his streams. 
In general, we should not be closing streams that we did not create ourselves, no?

Reply:
The reader is only closed for multipart/form-data requests since no other requests will have a non-empty list of file items and the input for those form requests are always from FileItem.getInputStream(), not the servlet input stream.